### ISSUE TITLE
Walking inclined paths

### ISSUE THREAD

**Issue by hungerburg** [valhalla/valhalla#3969]
There are several crude approximations for estimating time when walking non flat terrain.

E.g. for this walk from the train station to the hospital - https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.249715685844423%2C47.28104865174184%2C11.24449610710144%2C47.2835305523123 Valhalla estimates 9min20sec. Very sportive :) Walking speed set to 4km/h estimates 12 minutes, still not bad for this journey :)  For the record:

On https://en.wikipedia.org/wiki/Naismith%27s_rule Under Langmuir corrections it says: 4 km/h distance + 450 m/h of ascent, that makes

```
600 m => 1/40*6 => 0.15h
+ 1/450*90      => 0,2h
                   ----
                   0,35h => 21min

```
On https://www.alpenverein.at/portal/news/aktuelle_news_kurz/2018/2018_06_14_wie-berechnet-man-die-gehzeit-auf-wanderwegen.php it says: 300m/h ascent, 500m/h descent, 4km/h distance and a formula: The smaller value gets halved and added to the larger one:

```
90m elevation => 90/300   => 0.3h
600m distance => 1/400*60 => 0.15h
                             ----
0.3+0.15/2                => 0.375 => 22.5min
```

Estimates on guideposts usually have a granularity of quarters of an hour, so a guidepost there would read half an hour. This issue not a bug report, just an idea for enhancement. Though, if you care for hiking, this should get in, because wrong estimates on duration are in the top reasons for people calling emergency.

**Comment by kevinkreiser** [comment]
not too long ago we did try to modulate the speed of the pedestrian costing by the grade (slope) of a given edge in the graph: #3234 

have a read of that and see how we ended up implementing it. basically what we store in the graph is an "average" slope per graph edge and then we have a look up table (the slopes are broken into only 16 values from 'very much downhill' to 'flat' to 'very much uphill') which matches a given slope to a given multiplier that modifies ones walking speed.

in the end it is possible for us to measure total ascent but we decided not to do it that way and instead relied on a  weighted average slope value. its possible that those values could be modified but they did have some scientific grounding. at any rate, i wanted to let you know that at least we do try to modify the speed based on the ascent or descent we think you are going to be doing.

as you pointed out, i think its still possible for valhalla to overestimate the travel time for pedestrian routes. would you like to try your hand at modifying those values and testing the results to see if you can come up with a more balanced set of multipliers? we've done not exhaustive quality testing wrt pedestrian ETAs so thsi would be something the project could really use if you'd be willing. we can coach you on how to do it if you are interested.

thanks!

**Comment by hungerburg** [comment]
I skimmed the linked issue and the [article](https://www.myopencountry.com/average-hiking-speed/). Then I went to the the drawing board and noticed, that an angle of 5.5° happily can be expressed as an incline of 10%, much more conveniently.

I found a GPX of my whereabouts where I did the journey in top post in about 12 minutes without breaking a sweat including taking photos for OSM mapping. We will never learn the guidepost timings, because there wont ever be a guidepost with estimates there, as people simply are not supposed to take that route. I might even do it in 9:20 minutes given some dedication.

Below a plot of what the Alpine Club pretends to use to gauge guidepost timings, which I learned is taken from a DIN standard: The chart has some nice properties, I am not enough mathematically minded though to find a formula that turns speed into a function of incline. Theses properties are: At 7.5% incline (reminds of 3/4), elevation becomes the base for calculating duration. That base simply is double incline. This looks suspicious! People drawing up DIN standards sometimes get driven by such beauties ;)

![Marschzeiten](https://user-images.githubusercontent.com/586972/218886711-62c0f661-c14a-401a-8349-7f29ccbae8fd.png)

**Comment by nilsnolde** [comment]
I didn’t read through the DIN spec or article yet, but thanks for sharing that. If those diagrams are the main take-away, I agree, that’s odd haha I’d actually expect a ready-made formula.. 

Anyways, I think what you got there might be better already as what we got currently. It shouldn’t be too hard to make that a function. As it is, those slope values are much too spaced out to use them directly and interpolate “linearly” between them. Even though their slope/duration graph seems to suggest exactly that, I find that too few control points over a too big slope space. 

But that’s just my intuition. Would be good to compare the outcome for costing side by side with what we currently have for those slope values we know from the DIN spec.

**Comment by hungerburg** [comment]
@nilsnolde To make this clear, the chart is not from the DIN. I made it myself to get a picture of how their formula looks when plotted :) I could not get my hands on the DIN paper, but it likely contains just the verbal account. After all, it is about design of guideposts.

What you have already actually is not so bad, e.g. [from Rauschbrunnen to Hohe Warte](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.347133517265322%2C47.27906887044781%2C11.346441507339478%2C47.27964752742188%2C11.334843635559082%2C47.28403274090339%2C11.333599090576174%2C47.291572274726%2C11.333497166633606%2C47.30158087350774) it gives 3:17 hours, where the [guidepost](https://www.carto.net/andre.mw/photos/2008/10/05_hoetting_rauschbrunnen/20081005-145836_wegweiser_beim_gasthof_rauschbrunnen.shtml) says 4 hours. The DIN formula would estimate 1500/300+5/4/2=5.6 hours - which is overblown for people that do such hikes. I can only guess, destinations with a black dot might use different parameters. Yet, 5+ hours should match the performance of quite some of the guests to the Rauschbrunnen restaurant :)

PS: Walking speed set to 4km/h, valhalla estimates 4 hours! You still need to be trained to beat that.

**Comment by hungerburg** [comment]
The subject seems trickier than anticipated. Looking more closely at the [linked issue](https://github.com/valhalla/valhalla/pull/3234), and the commit to the valhalla codebase, I found this table kGradeBasedSpeedFactor purportedly supported by scientific evidence from https://mtntactical.com/research/yet-calculating-movement-uneven-terrain/ - My guess, the EF (efficiency factor) in German is called Leistungskilometer.

Still no idea, if tweaking the values of the table might be all it takes to give valhalla more reasonable estimates of journeys over inclined paths. There is no mention on how the factors have been calculated. The table seems to consider walking downhill speedier than walking flat terrain? This is not true on unpaved surfaces. The table stops at 15% incline/10% decline.

How to learn such subtleties from OSM data? 5 km/h is fine for paved footways in flat terrain. Unpaved paths rather go with 4 km/h in flat terrain. Change in altitude is under estimated.

My reason to meddle in this: Ever more people fire up some app and blindly believe, what it tells them.

**Comment by kevinkreiser** [comment]
you make a great point about not walking faster downhill. and yes I was hoping we would just plot your curve into the buckets with the multipliers and test the results

**Comment by hungerburg** [comment]
I was looking more for a formula, but [this table](https://github.com/valhalla/valhalla/blob/28889b6b1dab7888f713fea5435b09ddb9997b69/src/sif/pedestriancost.cc#L171) can be populated too.

```
$ ./DIN2grades.py 
> -30.0: 0.345
> -20.0: 0.476
> -15.0: 0.588
< -10.0: 0.714
<  -8.0: 0.758
<  -6.5: 0.794
<  -5.0: 0.833
<  -3.0: 0.893
<  -1.5: 0.943
<   0.0: 1.000
<   1.5: 0.909
<   3.0: 0.833
<   5.0: 0.750
<   6.5: 0.698
>   8.0: 0.638
>  10.0: 0.545
>  11.5: 0.492
>  13.0: 0.448
>  15.0: 0.400
>  20.0: 0.316
>  30.0: 0.222
```

Here script to generate the modifiers [DIN2grades.txt](https://github.com/valhalla/valhalla/files/10755521/DIN2grades.txt)

**Comment by hungerburg** [comment]
Still a bit uncertain, the DIN algorithm gives really high estimates, yet everywhere it is praised. So there must be something to it. Perhaps these timings are quite appropriate for a general purpose pedestrian router, that is not targeting avid hikers?

If tweaked, using 400m ascent per hour, it [revealed only a 35% decrease per 10% incline](https://mtntactical.com/research/walking-uphill-10-grade-cuts-speed-13not-12/) (quoted literally) instead of the anticipated halving :) (Should then also use 600m descent, but keep 4km/h.) That might be good for a hiking profile?

I used the trail to Hohe Warte, because this looks a good candidate to test incline. The route goes practically just straight up, only the last 50m or so needs some scrambling. The Swiss spreadsheet gives 4:05 for 1500 m of ascent over 5km distance, i.e. 30% incline, and 2:20 for descent. That might have had some influence on the guidepost at Rauschbrunnen. I found a gpx of mine, where ascent took me 3:30 including half an hour chat at the monument. I wager to say, there are local athletes that can do it in 1:30, maybe they even post on strava.

**Comment by nilsnolde** [comment]
Thanks @hungerburg! 

Do you have the time to quickly hack that table in the grade array of the pedestrian costing and see what it does for you? I appreciate we can't offer the full range that's interesting for hiking (i.e. only -10° - +15°) and there's still a few things missing, personally I think your surface-based modulation also makes sense, even for the more generic use of pedestrian. 

There's a chance we'll see that currently navigating hikers in demanding environments is not very viable due to the limited grade we can save for an edge. Not sure if that'd change once we get to implement #3944 .

**Comment by hungerburg** [comment]
I did now fork and commit https://github.com/valhalla/valhalla/compare/master...hungerburg:valhalla:master - unfortunately, I did not commit into a new branch. How to continue?

**Comment by nilsnolde** [comment]
You'd just have to build Valhalla now with your changes, build a graph for the Alps. Then clone this: https://github.com/gis-ops/valhalla-app, change the `.env`'s `REACT_APP_VALHALLA_URL` to `http://localhost:3000`, follow the installation instructions in the README and look at it in the browser.

Let me know if that's too much for you. I'm almost inclined to just put it on the public server haha.

**Comment by hungerburg** [comment]
@nilsnolde :) This is beyond my skills. Will github build it, add a green checkbox, can you pull it, do I have to push it?

**Comment by nilsnolde** [comment]
Don't worry, I'll really add it on the server. It's safe enough. Then we can all play around with your new values on https://valhalla.openstreetmap.de/.

I'll push your changes to a new branch here, it's easier for me to deploy.

**Comment by hungerburg** [comment]
If this ever proves useful, I am fine with whatever licence the project has and 'd be honoured to have contributed.

Looking forward to try out. No idea what will be affected. Urban should be safe?

**Comment by nilsnolde** [comment]
Alright, that's it. Can you try a few of your examples @hungerburg ?

**Comment by nilsnolde** [comment]
Once you're happy (we'll try a bit as well), you can PR your changes that you already did, with a bit more documentation in the code what the calculation is based on (similar to what's there now), and then we can merge.

**Comment by hungerburg** [comment]
Urban is not safe - The [Thaliastraße](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=16.337941288948063%2C48.20856998432531%2C16.306677460670475%2C48.21338880449155) just goes up continuously with a very gentle slope. Over 2.4km it rises 24m, i.e. 1% incline, it does not go up 41 and 16 down. The values from the exponential make it all smooth again. The values from DIN exaggerate the error in the terrain model.

**Comment by hungerburg** [comment]
Hello @nilsnolde , No idea what you did, these certainly are not DIN timings, but much to the better! Here some samples, with these settings {"use_hills":1,"walking_speed":5.1,"max_hiking_difficulty":6} All the descents shown in the directions pane are artefacts from the terrain model. On the ground the trails continuously go up.

[Hohe Warte 4:30](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.347128152847292%2C47.27907250981961%2C11.334961652755739%2C47.28419649702193%2C11.33368492126465%2C47.291310301226616%2C11.333486437797546%2C47.30158087350774), [Höttinger Alm 1:52](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.395182609558107%2C47.28471323522792%2C11.379529237747194%2C47.292209010460176%2C11.36686384677887%2C47.29847771926653), [Bodensteinalm 1:43](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=11.398251056671144%2C47.286419891041376%2C11.387801170349121%2C47.303388840714284) - Bodensteinalm should take longer than Höttinger Alm, same distance but steeper for longer sections, nearly 200m more altitude to climb.

Are [AllTrails timings](https://www.alltrails.com/de/route/austria/tyrol/von-hungerburg-zur-hottinger-alm) based on user provided GPX tracks? They say Höttinger Alm 3:52 up and down. Which reminds of DIN timing ;)

**Comment by nilsnolde** [comment]
> Urban is not safe - The [Thaliastraße](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=16.337941288948063%2C48.20856998432531%2C16.306677460670475%2C48.21338880449155) just goes up continuously with a very gentle slope.

Setting aside the elevation errors/artifacts (and I'm not entirely our web app is really doing the right thing here), I still find that quite ok in terms of estimated time, right?

> No idea what you did, these certainly are not DIN timings, but much to the better!

I didn't do anything, that was all your input;) We'll never get perfect, mostly due to noisy elevation data, but am I right reading that you're happier than before?

**Comment by hungerburg** [comment]
The estimates are excellent. Thaliastraße seems mostly just flat speed. The hiking routes match what is commonly stated in tour descriptions by non-athletes. I fear a bit, that when the steeper grades are taken into account (now it stops at +15?), timings might explode. But they did not do so now, even in trails without extra steep sections, I am not in a position to speculate why; so maybe just wait and reevaluate then?

**Comment by nilsnolde** [comment]
Great. Jep, I'd just merge what you have in your master and then see what happens:) You're ok with that too @kevinkreiser ?

Then, you just have to raise a PR to our master with your changes (and some updates in the comments how the current values were derived, data source etc, can be as brief as possible). Also a short entry to the CHANGELOG.md. Then at least your work is recognized in that you're a contributor to the project. Sounds good?

**Comment by kevinkreiser** [comment]
yep let's make a pr and give it a quick review I wanted to double check the slope bucket values

**Comment by hungerburg** [comment]
From what I can see comparing with guidepost timings, valhalla estimates, while lower, are very reasonable for inclined paths :+1:

**Comment by kevinkreiser** [comment]
@hungerburg thank you again for putting in the work on that. we really appreciate it!

**Comment by xlqian** [comment]
Hello everyone, by bumping Valhalla, we observed a considerable decrease in walking speed in connection to this issue, which is interesting and totally makes sense. By digging into the subject, we realized that even on a "slightly downhill" (0% to -10%), the speed factor is under 1, it's probably true for mountain hikers but quite astonishing for a city walker.

here is what we found on the Wikipedia, 
https://en.wikipedia.org/wiki/Naismith%27s_rule#/media/File:Hiking_speed.svg

Per Tobler's function, on a slight downhill, the speed should increase by a tiny amount, then decrease as the downhill slope goes steeper. 

I also plotted the data suggested by "DIN" and by Tobler's function so we can have visual feedback:

![image](https://github.com/valhalla/valhalla/assets/6093486/d8b0f337-867d-4d81-9227-a53d7dff0c02)

It seems a lot of work has been done to validate the DIN formula, in order not to invalidate the previous work, we'd love to suggest a slight modification so the speed factor could look like this:

![image](https://github.com/valhalla/valhalla/assets/6093486/9e55f2d0-ef82-4740-9051-b8781a88383e)
